File: src/io/serial_reader.py
import time
import threading
from PySide6.QtCore import Signal, QThread
import serial


class SerialReaderThread(QThread):
    data_received = Signal(bytes)
    error_occurred = Signal(str)

    def __init__(self, serial_port, frame_timeout=50):
        super().__init__()
        self.serial_port = serial_port
        self.running = False
        self.buffer = bytearray()
        self.last_receive_time = 0
        self.buffer_start_time = 0
        self._lock = threading.Lock()
        self.set_frame_timeout(frame_timeout)

    def set_frame_timeout(self, frame_timeout):
        self.frame_timeout = max(frame_timeout, 1) / 1000.0

    def _emit_buffer(self):
        if not self.buffer:
            return
        self.data_received.emit(bytes(self.buffer))
        self.buffer.clear()
        self.last_receive_time = 0
        self.buffer_start_time = 0

    def run(self):
        self.running = True
        while self.running:
            try:
                if not self.serial_port.is_open:
                    break

                if self.serial_port.in_waiting:
                    with self._lock:
                        data = self.serial_port.read(self.serial_port.in_waiting)
                    current_time = time.monotonic()

                    if self.buffer and (current_time - self.last_receive_time) > self.frame_timeout:
                        self._emit_buffer()

                    if not self.buffer:
                        self.buffer_start_time = current_time

                    self.buffer.extend(data)
                    self.last_receive_time = current_time

                    if (current_time - self.buffer_start_time) >= self.frame_timeout:
                        self._emit_buffer()
                else:
                    if self.buffer and self.last_receive_time > 0:
                        if (time.monotonic() - self.last_receive_time) > self.frame_timeout:
                            self._emit_buffer()

                    self.msleep(10)
            except serial.SerialException as e:
                self.error_occurred.emit(f"\u4e32\u53e3\u9519\u8bef: {str(e)}")
                break
            except Exception as e:
                if self.running:
                    self.error_occurred.emit(f"\u8bfb\u53d6\u9519\u8bef: {str(e)}")
                    self.msleep(10)

        self.running = False

    def stop(self):
        self.running = False
        with self._lock:
            self._emit_buffer()
        self.wait(1000)
        if self.isRunning():
            self.terminate()

File: src/io/serial_reader.cpp
#include "serial_reader.h"

#include <QSerialPort>

SerialReaderThread::SerialReaderThread(QSerialPort *serial, int frameTimeoutMs, QObject *parent)
    : QThread(parent), m_serial(serial)
{
    setFrameTimeout(frameTimeoutMs);
}

void SerialReaderThread::setFrameTimeout(int timeoutMs) {
    m_frameTimeoutSec = timeoutMs / 1000.0;
}

void SerialReaderThread::stop() {
    m_running = false;
    requestInterruption();
    emitBuffer();
    wait(1000);
}

void SerialReaderThread::run() {
    m_lastByteTimer.start();
    m_bufferStartTimer.start();

    while (m_running && !isInterruptionRequested()) {
        if (!m_serial->isOpen()) {
            msleep(50);
            continue;
        }

        // Read available data
        if (m_serial->waitForReadyRead(10)) {
            QByteArray data = m_serial->readAll();
            if (!data.isEmpty()) {
                QMutexLocker locker(&m_bufferMutex);
                m_buffer.append(data);
                m_lastByteTimer.start();

                if (!m_bufferActive) {
                    m_bufferActive = true;
                    m_bufferStartTimer.start();
                }
            }
        } else {
            // No data available, check frame timeout
            QMutexLocker locker(&m_bufferMutex);
            if (m_bufferActive && !m_buffer.isEmpty()) {
                double idleSec = m_lastByteTimer.elapsed() / 1000.0;
                double durationSec = m_bufferStartTimer.elapsed() / 1000.0;

                // Emit if idle gap exceeds timeout or continuous stream duration >= timeout
                if (idleSec >= m_frameTimeoutSec || durationSec >= m_frameTimeoutSec) {
                    emitBuffer();
                }
            }
            locker.unlock();
            msleep(10);
        }
    }

    emitBuffer();
}

void SerialReaderThread::emitBuffer() {
    QMutexLocker locker(&m_bufferMutex);
    if (!m_buffer.isEmpty()) {
        QByteArray data = m_buffer;
        m_buffer.clear();
        m_bufferActive = false;
        locker.unlock();
        emit dataReceived(data);
    }
}
